src/compiler/checker.ts

```diff
@@ -27160,6 +27160,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             case SyntaxKind.ElementAccessExpression:
                 // The resolvedSymbol property is initialized by checkPropertyAccess or checkElementAccess before we get here.
                 return isConstantReference((node as AccessExpression).expression) && isReadonlySymbol(getNodeLinks(node).resolvedSymbol || unknownSymbol);
+            case SyntaxKind.ObjectBindingPattern:
+            case SyntaxKind.ArrayBindingPattern:
+                return isBindingElement(node.parent) ? isConstantReference(node.parent.parent) : isVariableDeclaration(node.parent) && isVarConstLike(node.parent);

```

Comment:
an alternative solution for this could look smth like this:
```suggestion
                const rootDeclaration = getRootDeclaration(node.parent);
                return isVariableDeclaration(rootDeclaration) && isVarConstLike(rootDeclaration);
```

Reply:
I believe that this should work as well: [TS playground](https://www.typescriptlang.org/play?ts=5.4.0-dev.20231104#code/C4TwDgpgBAqgzhAigVwgJxAJQnZAbYAHgBUA+KAXgCgooAfKAbxttYEs4BlZAYx5zgAuKADMAhngQBuFqygATMcDHDkAO3kQRbNRHky5AXxYNmc2h258Bw4GlQHzi5cOKPahmVRHqewNgD2alDICCjoIAAUAJTC8EioGNi4BIRqyAC2AEbo5GZQaBDAyGjB+RZcvPxwQqISCAA0sgpKKiEaWjp6TR4yxlQ8QXDATC0uY2IAjA1QllU2s5XWNZNQhpQhYYlR0V4+an6BwcA4wJGME8LOUzNzy7V31XCrhnFbEcn4ROnZudFMLEGamGo2uV1aACZbksnsJHgIIWsNqEEhEYo4gSCLmCJgBmaFWWGLQkCXFIiibVEYdFUQFDEZiQokmobeErKAAMg5xPmNURXJ591xjjYIigkUZEGZcH+5QmkwAdMAAgBRAAeYCCEDU-gk6KgAHoDVA4AALAL4eRQAIAa2a1whStVGq1OrYet2huNZoteCttvtrVxTvVmt0bo9Ui9JvNlutdo8VH6QA) but the problem is that we are dealing with a pseudo-reference here. At this point, we can't easily check if the original symbol for which we are performing the narrowing is a non-reassigned parameter.

Reply:
Wish there was a 'const' modifier available for parameters and we could just check that!

Reply:
This should make that possible: https://github.com/microsoft/TypeScript/pull/56313